src/flb_input_chunk.c

```diff
@@ -1555,6 +1530,31 @@ static int input_chunk_append_raw(struct flb_input_instance *in,
         flb_free(filtered_data_buffer);
     }
 
+    if (ret == CIO_OK) {
+        ic->added_records =  n_records;
```

Comment:
Are `ic->added_records` and `ic->total_records` used in any code not guarded by `FLB_HAVE_METRICS`? If not, might as well move this update inside the guard…

Reply:
See my note in the PR description, but here's some more details.

Here, `event_chunk_create` is called with `total_records`, which is only set with `FLB_HAVE_METRICS`. https://github.com/fluent/fluent-bit/blob/8733b25df17b74fd3b2aa7594594451db79d1b25/src/flb_task.c#L370

But there are a number of output plugins outside of a `FLB_HAVE_METRICS` definition that seem to expect this number to be set on the event chunk.
https://github.com/fluent/fluent-bit/blob/8733b25df17b74fd3b2aa7594594451db79d1b25/plugins/out_stackdriver/stackdriver.c#L2751-L2754
https://github.com/fluent/fluent-bit/blob/8733b25df17b74fd3b2aa7594594451db79d1b25/plugins/out_loki/loki.c#L1545-L1550
https://github.com/fluent/fluent-bit/blob/8733b25df17b74fd3b2aa7594594451db79d1b25/plugins/out_datadog/datadog.c#L126

Reply:
Got it, makes sense. FWIW, the code in https://github.com/fluent/fluent-bit/blob/8733b25df17b74fd3b2aa7594594451db79d1b25/src/flb_task.c#L370 also seems buggy, since it'll only set `total_events` when `FLB_HAVE_METRICS` is true. Should we take this opportunity to remove that guard as well?

Reply:
After seeing this I realize that `ifdef` guard in `flb_task.c` also needs to be removed so I'll do that shortly.

Reply:
GitHub UI didn't update lol; yes I will remove that guard.

Reply:
Hi @braydonk, the original code updated the input plugin metrics before performing the filtering because it immediately wrote the data to the chunk file which is why it's done in that way, however, there's a bug in the conditional in line 1504 (original) because `ret` has not been initialized to the expected value which was my mistake when I rearranged it.

I was about to send a PR to fix that as well as the record rollover / drop counter which should retain the original flow, however, I think your test case is a good addition and since you have already opened this one I'd rather discuss it with you to  reach an agreement and merge your patch, would you like to do that?

Reply:
Sure, I'll message you offline to figure out what to change here

Reply:
Discussed on Slack, decided to keep this approach to the ensure correctness of the metrics being updated.